### PR TITLE
Use block sparse input for the first layer.

### PR DESCRIPTION
**Pull request by AndrovT**
Use block sparse input for the first fully connected layer on architectures with SSSE3. The net is the exact same as in #4611 except that the feature transform output is sorted by likelihood of that neuron being nonzero.

Passed STC:
https://tests.stockfishchess.org/tests/view/6485aa0965ffe077ca12409c
LLR: 2.93 (-2.94,2.94) <0.00,2.00>
Total: 8864 W: 2479 L: 2223 D: 4162
Ptnml(0-2): 13, 829, 2504, 1061, 25 

bench 2614824

### CODE REVIEW

**Comment by cj5716** [comment]
Perhaps adding bench speedup results in commit description would be good

**Comment by vondele** [comment]
@AndrovT  nice speedup. Locally i measure 13% speedup. 

As mentioned on discord, I think we need two things before we merge
a) ensure the results are correct on architectures we don't test well in CI (e.g. vnni or avx512, neon, etc)
b) have a link to the script used for permuting the network weights. Ideally we have that script integrated in the nnue-pytorch repository.

@Sopel97 any chance you could review this patch?

**Comment by Sopel97** [comment]
Nice! Good to see this approach finally paying off. Do you know what the average density of non-zero values is with the current nets?

> feature transform output is sorted by likelihood of that neuron being nonzero

Have you checked if this actually matters? The finding of NNZs uses a non-branching implementation based on https://github.com/syzygy1/Cfish/issues/204, which should not depend in any way on the order.

File: src/nnue/layers/affine_transform_sparse_input.h [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+#include "simd.h"
+
+/*
+  This file contains the definition for a fully connected layer (aka affine transform) with block sparse input.
+*/
+
+namespace Stockfish::Eval::NNUE::Layers {
+
+#if defined(USE_SSSE3)
+  const std::array<std::array<std::uint16_t, 8>, 256> lookup_indices = [](){
```
Comment (Sopel97) [inline src/nnue/layers/affine_transform_sparse_input.h:39]: These are potential ODR violations. Should probably be marked `static inline`. Also, they could be made cache aligned to improve worst case performance.

File: src/nnue/layers/affine_transform_sparse_input.h [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+    #define vec_nnz(a) _mm512_cmpgt_epi32_mask(a, _mm512_setzero_si512())
+#elif defined (USE_AVX2)
+    using vec_t = __m256i;
+    #define vec_nnz(a) _mm256_movemask_ps((__m256)_mm256_cmpgt_epi32(a, _mm256_setzero_si256()))
+#elif defined (USE_SSSE3)
+    using vec_t = __m128i;
+    #define vec_nnz(a) _mm_movemask_ps((__m128)_mm_cmpgt_epi32(a, _mm_setzero_si128()))
+#endif
+    constexpr IndexType ChunksPerOutput = std::max<IndexType>(8 * 4 / sizeof(vec_t), 1);
+    constexpr IndexType OutputsPerChunk = std::max<IndexType>(sizeof(vec_t) / 4 / 8, 1);
```
Comment (Sopel97) [inline src/nnue/layers/affine_transform_sparse_input.h:84]: These are confusingly named, appear to contradict.

File: src/nnue/layers/affine_transform_sparse_input.h [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+              j &= j - 1;
+              ++k;
+          }
+          v[i] = k;
+      }
+      return v;
+  }();
+
+  template<const IndexType InputDimensions>
+  void find_nnz(const std::int32_t* input, std::uint16_t* out, IndexType& count_out) {
```
Comment (Sopel97) [inline src/nnue/layers/affine_transform_sparse_input.h:72]: This function could use some comments, as it's not obvious what it does from the name. It gives back indices to non-zero inputs, but, crucially, it groups them in groups of 4, because later accumulation does 4 outputs at a time. So the first 4 inputs have index 0, the next 4 have index 1, and so on. Input chunk is considered non-zero if any of the 4 inputs in the group is non-zero.

File: src/nnue/layers/affine_transform_sparse_input.h [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+      const auto input32 = reinterpret_cast<const std::int32_t*>(input);
+
+      find_nnz<NumChunks>(input32, nnz, count);
+
+      const vec_t* biasvec = reinterpret_cast<const vec_t*>(biases);
+      vec_t acc[NumRegs];
+      for (IndexType k = 0; k < NumRegs; ++k)
+        acc[k] = biasvec[k];
+
+      for (IndexType j = 0; j < count; ++j)
```
Comment (Sopel97) [inline src/nnue/layers/affine_transform_sparse_input.h:219]: later, manual unrolling by 2x could be tried. Currently this loop has a lot of linear dependencies.

**Review by Sopel97 — CHANGES_REQUESTED** [review]
First, I strongly suggest testing this with the current master network, as the order of weights shouldn't matter.

Second, I cannot verify avx512/vnni512 right now, but the code looks correct. I'll add benches later

**Comment by MinetaS** [comment]
All tests are performed with bench options "1024 1 18 default depth nnue".
Binaries are built under PGO with Clang 15.0.7.

Processor:  Intel(R) Xeon(R) Platinum 8272CL CPU @ 2.60GHz

* x86-64-vnni512: 2.1% speedup
```
Result of 100 runs
==================
base (./stockfish-master) =    1449135  +/- 2915
test (./stockfish       ) =    1479613  +/- 1946
diff                      =     +30478  +/- 1998

speedup        = +0.0210
P(speedup > 0) =  1.0000
```

* x86-64-avx512: 0.6% speedup
```
Result of 100 runs
==================
base (./stockfish-master) =    1446929  +/- 2227
test (./stockfish       ) =    1455462  +/- 2387
diff                      =      +8533  +/- 1514

speedup        = +0.0059
P(speedup > 0) =  1.0000
```

* x86-64-vnni256: 7.2% speedup
```
Result of 100 runs
==================
base (./stockfish-master) =    1394096  +/- 1398
test (./stockfish       ) =    1495079  +/- 1684
diff                      =    +100984  +/- 1507

speedup        = +0.0724
P(speedup > 0) =  1.0000
```

* x86-64-bmi2: 4.1% speedup
```
Result of 100 runs
==================
base (./stockfish-master) =    1424134  +/- 1850
test (./stockfish       ) =    1482731  +/- 2060
diff                      =     +58596  +/- 2264

speedup        = +0.0411
P(speedup > 0) =  1.0000
```

* x86-64-sse41-popcnt: 6.4% speedup
```
Result of 100 runs
==================
base (./stockfish-master) =    1215782  +/- 1163
test (./stockfish       ) =    1293076  +/- 1204
diff                      =     +77294  +/- 1340

speedup        = +0.0636
P(speedup > 0) =  1.0000
```

* x86-64-ssse3: 6% speedup
```
Result of 100 runs
==================
base (./stockfish-master) =    1190862  +/- 1218
test (./stockfish       ) =    1262138  +/- 1190
diff                      =     +71276  +/- 1263

speedup        = +0.0599
P(speedup > 0) =  1.0000
```

Results of "bench 4096 1 30 default depth nnue":

| Architecture        | Signature |
|---------------------|-----------|
| x86-64-vnni512      | 442205461 |
| x86-64-avx512       | 442205461 |
| x86-64-vnni256      | 442205461 |
| x86-64-bmi2         | 442205461 |
| x86-64-sse41-popcnt | 442205461 |
| x86-64-ssse3        | 442205461 |

**Comment by vondele** [comment]
Here some numbers comparing 
* master
* patch (this PR, including the reordered net) 
* SameNet (this PR but keeping the original net)

TLDR: patch 13% speedup, SameNet 11% speedup, Net reorder benefit 2% (consistent with reports of this in Discord)

on `AMD Ryzen 9 3950X`

```
Result of  10 runs
==================
base (./stockfish.master       ) =    1394168  +/- 12402
test (./stockfish.patch        ) =    1576029  +/- 10992
diff                             =    +181860  +/- 4862

speedup        = +0.1304
P(speedup > 0) =  1.0000
```

```
Result of  10 runs
==================
base (./stockfish.master       ) =    1381918  +/- 4253
test (./stockfish.SameNet      ) =    1529336  +/- 5972
diff                             =    +147418  +/- 4775

speedup        = +0.1067
P(speedup > 0) =  1.0000
```

```
Result of  10 runs
==================
base (./stockfish.patch        ) =    1563712  +/- 10769
test (./stockfish.SameNet      ) =    1536426  +/- 10142
diff                             =     -27286  +/- 3684

speedup        = -0.0174
P(speedup > 0) =  0.0000
```

**Comment by Sopel97** [comment]
looks good to me now

**Comment by vondele** [comment]
For reference, the reordering of the net was described here https://discord.com/channels/435943710472011776/813919248455827515/1117464914341134428 :

Used https://github.com/AndrovT/Stockfish/tree/log-activations to find out the likelihood of activation of the neurons and this python script

```python
import torch

def permute_l1(nnue, permutation):
    l1_size = nnue.layer_stacks.l1.in_features
    assert l1_size == len(permutation)*2

    permutation.extend([x + l1_size // 2 for x in permutation])
    ft_permutation = permutation + list(range(l1_size, nnue.input.num_outputs))

    nnue.input.weight.data = nnue.input.weight.data[:, ft_permutation]
    nnue.input.bias.data = nnue.input.bias.data[ft_permutation]
    nnue.layer_stacks.l1.weight.data = nnue.layer_stacks.l1.weight.data[:, permutation]
```

Ideally this can be added in some way to the trainer. It might also be possible to improve the scheme with which the outputs are permuted (e.g. exploiting correlation between the outputs, not just their probability of being zero).

**Comment by vondele** [comment]
Thanks, nice first contribution!

**Comment by mstembera** [comment]
I wonder if this swings the balance toward even larger nets?

**Comment by Technologov** [comment]
> I wonder if this swings the balance toward even larger nets?

VizVez thinks the same. Great minds think alike !
